### PR TITLE
Add support for DataRequirements to the packager

### LINKED ISSUE #395: Add minimal DataRequirement support
Currently, our SDK packager produces Library resources without DataRequirements. This means that servers running such a library must be able to supply all data for the context (most often, a Patient) to the engine. This is wasteful, since Libraries will never use all data for a patient.

By generating basic DataRequirements specifying the resource type used in the `Retrieve` statements in the ELM, we can limit the set of data needed to fetch and pass on to the engine.

### PR DESCRIPTION
**Pull request by ewoutkramer**
Fixes https://github.com/FirelyTeam/firely-cql-sdk/issues/395.

Note that this really is a basic first version, just adding the names of the resource types as datarequirements to the Library when packaging.

### CODE REVIEW

**Comment by ewoutkramer** [comment]
Not that basic anymore, as I ported (part of) the public Java analyzer to C#, so we at least have codes to limit the retrieves as well.

**Comment by ewoutkramer** [comment]
This is an example of a DataRequirement that we now add to our Library resources:

```json
  {
      "type": "Encounter",
      "profile": [
        "http://hl7.org/fhir/StructureDefinition/Encounter"
      ],
      "mustSupport": [
        "type"
      ],
      "codeFilter": [
        {
          "path": "type",
          "valueSet": "https://www.ncqa.org/fhir/valueset/2.16.840.1.113883.3.464.1004.1761"
        }
      ]
    }
```

File: Cql/Cql.Compiler/LibrarySet.cs [hunk 1]
```diff
@@ -213,11 +217,14 @@ private void RecalculateStateIfNecessary()
     /// <param name="lib">The name of the library to load.</param>
     /// <param name="version">The version of the library to load.</param>
     /// <returns>A collection of loaded libraries, including the specified library and its dependencies.</returns>
+    /// <remarks>Supply the name of the library in <paramref name="lib"/>, not the name of the file.</remarks>
     public IReadOnlyCollection<Library> LoadLibraryAndDependencies(
         DirectoryInfo elmDirectory,
         string lib,
         string version = "")
     {
+        if(lib.EndsWith(".cql")) lib = lib[..^4];
```
Comment (baseTwo) [inline Cql/Cql.Compiler/LibrarySet.cs]: This should be case-insensitive comparison

File: Cql/Cql.Compiler/LibrarySetDefinitionCache.cs [hunk 2]
```diff
@@ -0,0 +1,108 @@
+using System;
```
Comment (baseTwo) [inline Cql/Cql.Compiler/LibrarySetDefinitionCache.cs:9]: Missing file header

File: Cql/Elm/IDefinitionElement.cs [hunk 3]
```diff
@@ -296,9 +308,10 @@ public partial class UsingDef : IDefinitionElement
     partial class IdentifierRef : IDefinitionElement
     {
         [JsonIgnore]
-        public AccessModifier Access => AccessModifier.Private;
+        AccessModifier IDefinitionElement.Access => AccessModifier.Private;
```
Comment (baseTwo) [inline Cql/Elm/IDefinitionElement.cs:311]: Is JsonIgnore still necessary on explicit interface properties?

**Review by baseTwo — CHANGES_REQUESTED** [review]

File: Cql/CoreTests/Elm/TestDataRequirementsAnalyzer.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
+using System.IO;
```
Comment (baseTwo) [inline Cql/CoreTests/Elm/TestDataRequirementsAnalyzer.cs:1]: NEeds file header

File: Cql/CoreTests/Elm/TestElmResolution.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
+using System.IO;
```
Comment (baseTwo) [inline Cql/CoreTests/Elm/TestElmResolution.cs:1]: NEeds file header

File: Cql/Cql.Compiler/Errors.cs [hunk 6]
```diff
@@ -17,16 +17,38 @@ namespace Hl7.Cql.Compiler;
...
+    public string GetMessage() => $"Library has an {Def.GetType()} element without a name. Library Identifier: '{Library}', Name: '{Def.Name}'.";
+}
+
 internal readonly record struct LibraryIdentifierMustBeUniqueError(Library Library) : ILibraryError
 {
     public string GetMessage() => $"Library did not have a unique name and version in the set. Duplicate Identifier: '{Library}'";
 }
 
-internal readonly record struct LibraryAliasUnresolvedError(Library Library) : ILibraryError
+internal readonly record struct LibraryDoesNotAppearAsADependencyError(Library Library) : ILibraryError
```
Comment (baseTwo) [inline Cql/Cql.Compiler/Errors.cs]: The message returned doesn't match with the name of this method

File: Cql/Cql.Compiler/LibrarySet.cs [hunk 7]
```diff
@@ -210,16 +214,20 @@ private void RecalculateStateIfNecessary()
...
     /// <param name="elmDirectory">The directory containing the ELM files.</param>
-    /// <param name="lib">The name of the library to load.</param>
+    /// <param name="libraryPath">The name of the library to load.</param>
     /// <param name="version">The version of the library to load.</param>
     /// <returns>A collection of loaded libraries, including the specified library and its dependencies.</returns>
+    /// <remarks>Supply the name of the library in <paramref name="libraryPath"/>, not the name of the file.</remarks>
     public IReadOnlyCollection<Library> LoadLibraryAndDependencies(
         DirectoryInfo elmDirectory,
-        string lib,
+        string libraryPath,
```
Comment (baseTwo) [inline Cql/Cql.Compiler/LibrarySet.cs]: libraryName could be a better name for this parameter. It is not the whole path of it that is passed in here

File: Cql/Cql.Compiler/LibrarySetDefinitionCache.cs [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        private LibraryCache()
+        {
+            // Nothing
+        }
+
+        private readonly Dictionary<Type, object> _cache = new();
+
+        private IDictionary<string, T> GetIndexForType<T>()
+        {
+            if (_cache.TryGetValue(typeof(T), out var index)) return (IDictionary<string, T>)index;
```
Comment (baseTwo) [inline Cql/Cql.Compiler/LibrarySetDefinitionCache.cs]: Put return on a separate line, then add another blank line after it. This improves the readability of the method exit point

File: Cql/Cql.Compiler/LibrarySetDefinitionCache.cs [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+            _cache.Add(typeof(T), newIndex);
+            return newIndex;
+        }
+
+        private static void Index<T>(
+            IDictionary<string, T> result,
+            Library library,
+            T[]? definitions) where T : IDefinitionElement
+        {
+            if (definitions is not { Length: > 0 }) return;
```
Comment (baseTwo) [inline Cql/Cql.Compiler/LibrarySetDefinitionCache.cs]: Return on its own line

File: Cql/Cql.Packaging/DataRequirementsAnalyzer.cs [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+    {
+        // Function toDataRequirement in the above Java file
+        if (node is not Elm.Retrieve retrieve) return false;
+
+        var dr = new DataRequirement();
+
+        // Get the resource type the requirement is for.
+        var (_, resourceType) = retrieve.dataType;
+        var knownType = Hl7.Fhir.Model.ModelInfo.FhirTypeNameToFhirType(resourceType);
+        if (knownType is null) return true; // Not a known FHIR type, so we can't do anything with it.
```
Comment (baseTwo) [inline Cql/Cql.Packaging/DataRequirementsAnalyzer.cs]: REturn on own line, followed by blank line

**Review by baseTwo — CHANGES_REQUESTED** [review]

File: Cql/Cql.Compiler/LibrarySetDefinitionCache.cs (hunk 9 above)
Comment (ewoutkramer) [reply Cql/Cql.Compiler/LibrarySetDefinitionCache.cs]: This is going to be hard, I have never done that in 30 years....

File: Cql/CoreTests/Elm/TestDataRequirementsAnalyzer.cs (hunk 4 above)
Comment (ewoutkramer) [reply Cql/CoreTests/Elm/TestDataRequirementsAnalyzer.cs:1]: Sigh...one thing I miss about VS.....splitting a class off into another file did copy the header in VS, but not in Rider. Oh well.....

Comment (ewoutkramer) [reply Cql/CoreTests/Elm/TestDataRequirementsAnalyzer.cs:1]: I noticed *none* (well, except one or two) of our test files have headers, so I won't fix that here. If we want our tests with headers, we'll need to make a PR to fix it for all of them, it does not make sense to just fix up a few.

Comment (baseTwo) [reply Cql/CoreTests/Elm/TestDataRequirementsAnalyzer.cs:1]: I guess it's just easier to remove the header from the unit tests, I don't think they would be useful to anyone as-is. [I'll create a ticket for that ](https://github.com/FirelyTeam/firely-cql-sdk/issues/436)

**Review by baseTwo — APPROVED** [review]

**Pull request merged**